Add new fields for multi role support.

ANeumann82 opened the pull request:
Summary:
- Add optional role field to App, Pod in API
- Add required role field in AppDefinition, PodDefinition
- Validate role field based on top-level-group enforceRole-flag
- Migration for Apps and Pods, setting default role

JIRA issues: MARATHON-8637, MARATHON-8638, MARATHON-8640

Co-authored-by: Karsten Jeschkies <kjeschkies@mesosphere.io>

jeschkies commented on src/main/scala/mesosphere/marathon/MarathonSchedulerService.scala:
```diff
@@ -267,14 +267,18 @@ class MarathonSchedulerService @Inject() (
-    Await.result(groupManager.invalidateGroupCache(), Duration.Inf)
+    //    logger.info("RefreshCaches before migration")
+    //    // refresh group repository cache
+    //    Await.result(groupManager.invalidateGroupCache(), Duration.Inf)
```
Maybe this should just be an invalidation. We probably want to adapt the comment above or verify that its claim is still true.

ANeumann82 replied:
I've cleaned that up and updated the comment.

jeschkies commented on src/main/scala/mesosphere/marathon/api/v2/AppHelpers.scala:
```diff
@@ -43,8 +44,11 @@ object AppHelpers {
       .unreachableStrategy.map(Raml.fromRaml(_))
       .getOrElse(UnreachableStrategy.default(hasPersistentVols))
+
+    val role = update.role.orNull
```
is `orNull` correct? Shouldn't we set the role from the group depending on the config?

EDIT: I had to expand the method. I'm a little confused. What is this method doing?

ANeumann82 replied:
Yeah, I think it is correct, although it looks really weird. 
This method converts an AppUpdate to an App, by converting it to an AppDefinition and than back to an App.... Having a null value in the AppDefinition leads to an Option.empty in the resulting App.
I've added a comment though.

jeschkies commented on src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala:
```diff
@@ -48,17 +50,24 @@ class AppsResource @Inject() (
+    val roleSettings = RoleSettings.forService(config, pathId, groupManager.rootGroup())
+    val normalizationConfig = AppNormalization.Configuration(config, roleSettings)
+    implicit val normalizer: Normalization[raml.App] = appNormalization(normalizationConfig)(AppNormalization.withCanonizedIds())
+    implicit val validator: Validator[AppDefinition] = AppDefinition.validAppDefinition(config.availableFeatures, roleSettings)(pluginManager)
```
There is no need to make it `implicit` since you are returning them explicitly anyways.

ANeumann82 replied:
True, was left over from C&P. Fixed

jeschkies commented on src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala:
```diff
@@ -48,17 +50,24 @@ class AppsResource @Inject() (
+    val roleSettings = RoleSettings.forService(config, pathId, groupManager.rootGroup())
+    val normalizationConfig = AppNormalization.Configuration(config, roleSettings)
+    implicit val normalizer: Normalization[raml.AppUpdate] = appUpdateNormalization(normalizationConfig)(AppNormalization.withCanonizedIds())
+    implicit val validator: Validator[AppDefinition] = AppDefinition.validAppDefinition(config.availableFeatures, roleSettings)(pluginManager)
```
See above.

ANeumann82 replied:
Fixed

jeschkies commented on src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala:
```diff
@@ -89,9 +98,13 @@ class AppsResource @Inject() (
       val now = clock.now()
-      val app = validateOrThrow(rawApp).copy(versionInfo = VersionInfo.OnlyVersion(now))
+
+      val app = validateOrThrow(Raml.fromRaml(ramlApp.normalize)).copy(versionInfo = VersionInfo.OnlyVersion(now))
```
I wonder if it should just be `normalizer.normalized(ramlApp)`. This ways `normalize` does not have to be implicit.

ANeumann82 replied:
I don't really like the pattern for the Normalizer, but in this case i'd leave it as is, both validator and normalizer are implicit and it is kind of clean. Making only the normalizer explicit wouldn't really help much

jeschkies replied:
Fair point,

jeschkies commented on src/main/scala/mesosphere/marathon/api/v2/validation/PodsValidation.scala:
```diff
@@ -250,11 +248,30 @@ trait PodsValidation extends GeneralPurposeCombinators {
+
+  private def validWithRoleEnforcement(roleSettings: RoleSettings): Validator[Pod] = validator[Pod] { pod =>
+    if (roleSettings.enforceRole) {
+      pod.role must notEmpty
```
This reads a little odd. I don't think that `enforceRole` is about requiring the role to be set but that the role must be the one from the top-level group. @timcharper could you clarify?

ANeumann82 replied:
Yeah, this was because the validation was done on the PodRaml, and before normalization was done. I've now changed it to validate the role only on the PodDefinition and after normalization.
At some point it might be a good idea to do validation/normalization with the same pattern on apps and pods...

jeschkies commented on src/main/scala/mesosphere/marathon/state/AppDefinition.scala:
```diff
@@ -465,7 +473,7 @@ object AppDefinition extends GeneralPurposeCombinators {
 
   def fromProto(proto: Protos.ServiceDefinition): AppDefinition =
-    AppDefinition(id = DefaultId).mergeFromProto(proto)
+    AppDefinition(id = DefaultId, role = null).mergeFromProto(proto)
```
Should we comment why `null` is fine here?

ANeumann82 replied:
Good idea.

jeschkies commented on src/main/scala/mesosphere/marathon/state/ResourceRole.scala:
```diff
@@ -8,6 +8,14 @@ import mesosphere.marathon.api.v2.Validation._
+    isTrue(s"""acceptedResourceRoles must be either [*] or one of ${validRoles.mkString("[", ",", "]")}""") { acceptedResourceRoles =>
+      acceptedResourceRoles.isEmpty ||
+        (acceptedResourceRoles.size == 1 &&
+          (acceptedResourceRoles.head.equals(Unreserved) || validRoles.contains(acceptedResourceRoles.head)))
```
I'm a little confused. is `acceptedResourceRoles` only allowed to be empty or size one? Why is that?

ANeumann82 replied:
After rereading the specs, i'm not totally sure. It seems that acceptedResourceRoles are actually not really compatible with custom roles? @timcharper, any clarifications?

ANeumann82 replied:
I'm still missing something here. Assuming we have a runSpec with `acceptedResourceRoles: ["slave_public"]` and marathon starts with `--mesos_role *` (or without --mesos_role).
After the migration for apps and pods is done, the runSpec would have `role: "*"`, which would then be invalid for `acceptedResourceRoles: ["slave_public"]`

Do we allow any acceptedResource roles if the role is set to "*"?

timcharper replied:
If `--mesos_role` is `*`, then acceptedResourceRoles can only be `"*"`

timcharper replied:
(I sure hope this is validated... lol)

ANeumann82 replied:
Hmmm. Then the default for `--mesos_role` should probably be 'slave_public', not '*', right? Otherwise I would assume we would break a lot of existing setups, as usually marathon is started without `--mesos_role` and users are setting `acceptedResourceRoles` to `slave_public` to assign the service to public facing nodes?

meichstedt replied:
This is **not** validated.

ANeumann82 replied:
Ok, i've updated the acceptedResourceRoleValidation to only accept the runSpec.role and/or *

timcharper replied:
Great, this is the correct behavior.

And.... shoot, it's not validated right now. Let's go forward with it validated for now. We'll likely need to implement a feature to fix nosensical resource roles, on-they-fly, and then have a deprecated-feature option to automatically fix them when they're posted erroneously

ANeumann82 commented on src/main/scala/mesosphere/marathon/util/RoleSettings.scala:
```diff
@@ -0,0 +1,62 @@
+    * @return A data set with valid roles, default role and a flag if the role should be enforced
+    */
+  def forService(config: MarathonConf, servicePathId: PathId, rootGroup: RootGroup): RoleSettings = {
+    val defaultRole = config.mesosRole()
```
@timcharper could you please review this method? It calculates the allowed roles for a given service...

jeschkies commented on src/main/scala/mesosphere/marathon/util/RoleSettings.scala:
```diff
@@ -0,0 +1,60 @@
+        RoleSettings(enforceRole = false, validRoles = Set(defaultRole), defaultRole = defaultRole)
+      } else {
+        val group = topLevelGroup.get
+        val defaultForEnforceFromConfig = false // TODO: Use value from config instead
```
I guess this relates to #6959. I don't think `defaultEnforceGroupRole` defines the default value but rather whether it's `off` or `top`.

ANeumann82 commented on src/main/scala/mesosphere/marathon/util/RoleSettings.scala:
```diff
@@ -0,0 +1,60 @@
+  * @param validRoles List of valid roles
+  * @param defaultRole The default role to use if no role is specified on the service. Defaults to '*'
+  */
+case class RoleSettings(enforceRole: Boolean, validRoles: Set[String], defaultRole: String) {
```
Good idea...

ANeumann82 commented on src/main/scala/mesosphere/marathon/util/RoleSettings.scala:
```diff
@@ -0,0 +1,60 @@
+      RoleSettings(enforceRole = false, validRoles = Set(defaultRole), defaultRole = defaultRole)
+    } else {
+      val topLevelGroupPath = servicePathId.rootPath
+      val topLevelGroup = rootGroup.group(topLevelGroupPath)
```
Well, there is the case with the group-endpoint where a full new group including apps is created. I haven't fully checked that, but at the normalization/validation stage we don't have the top-level groups created.

ANeumann82 replied:
As we've discussed, this is an edge case that should be fixed, but would probably require quite a bit of rework. I've added a TODO and log a warning if we encounter this case for now.

ANeumann82 commented on src/main/scala/mesosphere/marathon/util/RoleSettings.scala:
```diff
@@ -0,0 +1,60 @@
+      } else {
+        val group = topLevelGroup.get
+        val defaultForEnforceFromConfig = false // TODO: Use value from config instead
+        val enforceRole = group.enforceRole.getOrElse(defaultForEnforceFromConfig)
```
Well, if we only apply it to newly created groups, we probably should have the flag as not-optional and migrate it. If we have it as optional, this would probably be the best option, i guess?

jeschkies commented on src/main/scala/mesosphere/marathon/util/RoleSettings.scala:
```diff
@@ -0,0 +1,60 @@
+
+        if (enforceRole) {
+          // With enforceRole, we only allow the top-level group role
+          RoleSettings(enforceRole = true, validRoles = Set(group.id.root), defaultRole = group.id.root)
```
The API allows an empty `enforceRole` field, right? So we have to handle this case.

ANeumann82 replied:
At least in the RoleSettings we don't need the enforceRole-Parameter anymore, i've removed that.

ANeumann82 commented on src/test/scala/mesosphere/marathon/api/v2/validation/PodsValidationTest.scala:
```diff
@@ -206,13 +208,42 @@ class PodsValidationTest extends UnitTest with ValidationTestLike with PodsValid
+  "roleValidation" should {
+
+    "be invalid without role and role setting" in new Fixture {
+      val podDef = podWithoutRole.copy(role = Some("dev")).fromRaml.copy(role = null)
```
Removed it, was a left over from earlier attempts

ANeumann82 commented on src/test/scala/mesosphere/marathon/api/v2/validation/PodsValidationTest.scala:
```diff
@@ -206,13 +208,42 @@ class PodsValidationTest extends UnitTest with ValidationTestLike with PodsValid
+      )
+    }
+
+  }
```
Added some tests

timcharper commented:
@ANeumann82 - Everything is looking very good. Please see to it that the acceptance criteria that I've [specified in the ticket MARATHON-8638](https://jira.mesosphere.com/browse/MARATHON-8638) is implemented correctly. I'm not rejecting since I will be pretty out of commission tomorrow and don't want to hold up the merge. Specifically, needing attention is:

* Group-role should be allowed even if enforce_role is false. (currently, only the default role is allowed)
* The story around allowing existing roles to persist remains unaddressed (I think you're working on this?)

I think once those two items are done this will be good to merge.

ANeumann82 commented:
> @ANeumann82 - Everything is looking very good. Please see to it that the acceptance criteria that I've [specified in the ticket MARATHON-8638](https://jira.mesosphere.com/browse/MARATHON-8638) is implemented correctly. I'm not rejecting since I will be pretty out of commission tomorrow and don't want to hold up the merge. Specifically, needing attention is:
> 
> * Group-role should be allowed even if enforce_role is false. (currently, only the default role is allowed)
> * The story around allowing existing roles to persist remains unaddressed (I think you're working on this?)
> 
> I think once those two items are done this will be good to merge.

@timcharper I've cleaned up the RoleSetting calculation, now it's (hopefully) a bit more obvious what's happening. There was one bug if we were to create a pod/app in a non-existing group then the group role was not allowed - I've fixed that and added a unit test

Existing Roles should also be ok, there are unit tests for that.

We should validate the acceptance criteria with integration tests, i'll try to see if I can do something there, but I don't think that should stop us from landing
